Implement UID file object backup files and recovery processing.

simonqhughes opened the pull request:
The following provides more information on this commit:
- UID data is stored in object files. To improve robustness against power
  failure, a copy of the xxxx.dat file is stored in a xxxx.bak file. In the
  case that the xxxx.dat file is lost and an xxxx.bak file remains, the
  xxxx.dat file is restored from the backup file.
- The backup file and recovery algorithm are described in detail in the
  PSA trusted storage Low Level Design (LLD) markdown file.
- Test cases have been implemented in the psa-storage-example-app to verify the
  correct operation of the recovery processing code. The test cases are
  described in the LLD.

simonqhughes committed:
Implement UID file object backup files and recovery processing.

The following provides more information on this commit:
- UID data is stored in object files. To improve robustness against power
  failure, a copy of the xxxx.dat file is stored in a xxxx.bak file. In the
  case that the xxxx.dat file is lost and an xxxx.bak file remains, the
  xxxx.dat file is restored from the backup file.
- The backup file and recovery algorithm are described in detail in the
  PSA trusted storage Low Level Design (LLD) markdown file.
- Test cases have been implemented in the psa-storage-example-app to verify the
  correct operation of the recovery processing code. The test cases are
  described in the LLD.

simonqhughes committed:
Add PSA trusted storage low level design markdown file.

jainvikas8 commented on inc/config.h:
```diff
@@ -63,5 +70,13 @@
+/**
+ * \def PSA_STORAGE_TEST
+ *
+ * Define the this symbol to build a module test code into the library
```
Define this*

simonqhughes replied:
Fixed.

jainvikas8 commented on docs/psa_trusted_storage_lld.md:
```diff
@@ -0,0 +1,298 @@
+### Robustness Against Power Failure (RAPF)
+
+
+The current design is not Robust Against Power Failure (RAPF) because the Linux filesystem underlying ecryptfs (i.e. ext4) is susceptible to being left in an inconsistent state. This is because the filesystem is not transactional i.e. the filesystem metadata on the backing store requires more than 1 write operation to go from one self consistent state to the next. This can, for example, lead to t ...
```
Tools like fsck exist to restore filesystem metadata consistency when the corruption of these data structures takes place.

simonqhughes replied:
I dont understand your comment here. Is there are spelling or type mistake?

jainvikas8 commented on docs/psa_trusted_storage_lld.md:
```diff
@@ -0,0 +1,298 @@
+
+The starting conditions are as follows:
+
+1.    A file object exists called file1.dat, which has a backup file called file1.bak.x where x={0,1}. file1.tmp and file1.bak.x have identical object data.
```
Bullets and numbering, please.

simonqhughes replied:
If you look at the .md file in a browser you'll see the multiple "1." are rendered as an ordered list. Do this resolve your request?

simonqhughes replied:
psa_truste_storage_lld.md has now been removed.

jainvikas8 commented on docs/psa_trusted_storage_lld.md:
```diff
@@ -0,0 +1,298 @@
+In summary, the recovery process attempts to restore the latest version of a file object data, and maintain a backup copy by:
+
+1. Recovering any missing xxxx.dat.
+    1. If xxxx.bkx(new) exists then use it to recreate xxxx.dat using the modification procedure.
```
It would be good to mention - xxxx.bak.yyyy (new) and xxxx.bak.xxxx (old)?

simonqhughes replied:
Not sure exactly what you mean here. Are you saying that I better define terms? Are you saying:
- rather than referring to `xxxx.bak.yyyy (new)` in the text it would be better to use `xxxx.bak.yyy+1` ? 
- rather than referring to `xxxx.bak.yyyy (old)` in the text it would be better to use `xxxx.bak.yyy` ?

simonqhughes replied:
@jainvikas8 In conversation you asked me to fix this document so it uses a consistent naming convention for files when describing the procedure for restoring files from backups. AS discussed, the plan is to remove the lld from the PR. However, I've updated the HLD https://confluence.arm.com/display/mbedlinux/PSA+Storage+Software+HLD in the section "Robustness Against Power Failure (RAPF)" to use consistent naming. The LLD documentation was taken from here.

simonqhughes committed:
Remove docs/psa_trusted_storage_lld.md

simonqhughes committed:
Fix typo in inc/config.h.

jainvikas8 commented on lib/common_storage.c:
```diff
@@ -10,7 +10,11 @@
  *
  * This file is a modified version of mbed-crypto/library/psa_its_file.c v1.1.0d0.
+ *
+ * See the PSA Trusted Storage Linux Low Level Design docmment in the docs subdirectory
```
I think the statement would not make any sense, as the documentation is removed?

simonqhughes replied:
Removed.

jainvikas8 commented on lib/common_storage.c:
```diff
@@ -146,224 +503,2779 @@ static psa_status_t psa_its_read_file( psa_storage_uid_t uid,
+    status = psa_cs_init();
+    psa_assert( status == PSA_SUCCESS );
+    /* _init() should have recovered 1 file */
+    psa_assert( psa_cs_num_file_objects == 3 );
```
const define, seeing this used most of the places?

simonqhughes replied:
Added define.

jainvikas8 commented on lib/common_storage.c:
```diff
@@ -49,23 +68,73 @@
+    ( sizeof( PSA_CS_PREFIX ) - 1 +             /* prefix without terminating 0*/           \
+      sizeof( PSA_CS_ITS_SUBPREFIX ) - 1 +      /* sub-prefix without terminating 0*/       \
+      PSA_UID_STRING_LENGTH +                   /* UID (64-bit number in hex)*/             \
+      3 +                                       /* "_" and 8-bit sequence number */         \
```
Would it be suitable to make it a constant/define if it is used in many places?

simonqhughes replied:
Added define.

simonqhughes committed:
SQUASH: Fix review comments.

The following provides more information on this commit:
- Remove comment referring to PSA Trusted Storage Linux Low Level Design
  docmment in common_storage.c, as the document has been removed.
- Add define for length of sequence number string used in file names.
- Add define for checking the number of instantiated file objects.

simonqhughes commented:
Further review comments from Vikas from Simon/Vikas review meeting 20191107:
- Correct  PSA Stroage HLD RAPF picture:
  - correct (3) <-> (4)
  - add (6) for deleting old .bak file
  - remove seqnum and use yyy, and clearly defined that yyy is the sequence number
  - remove (yyy) from being appended to filenames and show them in the files.
- The `<uid>.tmp` file can be used for recovery under the following conditions:
    - A `<uid>.tmp(yyy+1)` file exists (with a seqnum yyy+1) but there is no `<uid>_yyy+1.bak` and no `<uid>.dat(yyy+1)`. (if the `<uid>_yyy+1.bak` then it can be used for recovery)
         This state can be detected by one of the following states:
             - a. A `<uid>_yyy.bak` file exists, or
             - b. A `<uid>.dat(yyy)` file exists.
         If the fact that that the `<uid>.tmp` sequence number has been updated indicates then `<uid>.tmp` file can be used to create `<uid>.dat(yyy+1)` in the absence of this file and the `<uid>_yyy+1.bak`
- Is there the possibility of that the `<uid>_yyy+1.bak` file can be left in a damaged half written state due to power failure?
- Where are what the test cases do documented?
- As part of the output of running the test cases, output a description of what the test case does.
- Make sure _DEBUG flag and the _TEST flag are independent.
- Add documentation about the methodology of how to add new test cases.
- Debug trace should output start/end of a particular test.

simonqhughes commented:
With respect to the 20191107 feedback from Vikas:
1. Item 1) has been addressed by updating the RAPF picture are requested. See the latest version of the HLD here: https://confluence.arm.com/display/mbedlinux/PSA+Storage+Software+HLD
2. It is possible to use the <uid>.tmp file for recovery under the described scenario. However, this hasn't been done in the current work. I've created a jira ticket for implementing this improvement here:

simonqhughes committed:
Fix robustness back-up file test cases output.

simonqhughes commented:
The following is a review question from Vikas:
>> Is there the possibility of that the <uid>_yyy+1.bak file can be left in a damaged half written state due to power failure?
It is not possible for the .bak file to be left in a damaged half written state because the full file is written to temporary file. Once written the file is then renamed to the .bak file with the POSIX rename() function, which is an atomic filesystem operation. Therefore the <uid>_yyy+1.bak either does not exist, or exists in its complete form.

The above

simonqhughes committed:
Delimit start/end of test cases in debug trace output.

simonqhughes commented:
@jainvikas8 I've worked through your feedback and acted upon it. Please review again.

jainvikas8 commented on lib/common_storage.c:
```diff
@@ -76,67 +143,356 @@
+
+
+/* FUNCTION: psa_cs_get_mktemp_filename
+ *  Generate a temporary filename using using an internal counter.
```
Typo/Grammer error

simonqhughes replied:
Fixed.

jainvikas8 commented on lib/common_storage.c:
```diff
@@ -146,224 +502,2828 @@ static psa_status_t psa_its_read_file( psa_storage_uid_t uid,
+        goto err0;
+    }
+
+    /* As all files are stored on encrypted file system, a request for no confidentiality
```
Typo/Grammer

simonqhughes replied:
Fixed.

jainvikas8 commented on lib/common_storage.c:
```diff
@@ -146,224 +502,2828 @@ static psa_status_t psa_its_read_file( psa_storage_uid_t uid,
+    status = psa_cs_read_file( uid, &info, &stream, api, &seqnum );
+    if( status == PSA_SUCCESS )
+    {
+        /* Step 1: Copy to pre-existing file to a tmp file.
```
Typo/Grammer

simonqhughes replied:
Fixed.

jainvikas8 commented on lib/common_storage.c:
```diff
@@ -146,224 +502,2828 @@ static psa_status_t psa_its_read_file( psa_storage_uid_t uid,
+    uint32_t get_filename_flags = PSA_CS_GET_FILENAME_F_NONE;
+    psa_status_t status = PSA_ERROR_GENERIC_ERROR;
+
+    /* NB: have to decrement sequence number supplied to _set() because creation of bak files will be
```
Typo/Grammer

simonqhughes replied:
Fixed.

jainvikas8 commented on lib/common_storage.c:
```diff
@@ -146,224 +502,2828 @@ static psa_status_t psa_its_read_file( psa_storage_uid_t uid,
+    uint32_t get_filename_flags = PSA_CS_GET_FILENAME_F_NONE;
+    psa_status_t status = PSA_ERROR_GENERIC_ERROR;
+
+    /* NB: have to decrement sequence number supplied to _set() because creation of bak files will be
```
Typo/Grammer

simonqhughes replied:
Fixed.

jainvikas8 commented on lib/common_storage.c:
```diff
@@ -146,224 +502,2828 @@ static psa_status_t psa_its_read_file( psa_storage_uid_t uid,
+
+
+/* FUNCTION: psa_cs_test_case_init()
+ *  Create a other uid files so there are present during recovery processing
```
Typo/Grammer

simonqhughes replied:
Fixed.

jainvikas8 commented on lib/common_storage.c:
```diff
@@ -146,224 +502,2828 @@ static psa_status_t psa_its_read_file( psa_storage_uid_t uid,
+    psa_status_t status = PSA_ERROR_GENERIC_ERROR;
+
+    /* remove uid file objects for uid1 and uid3.*/
+    status = psa_cs_remove( PSA_CS_TEST_UID1, state->api );
```
Could it be done in a loop with 2 different UID?

simonqhughes replied:
Fixed.

jainvikas8 commented on lib/common_storage.c:
```diff
@@ -146,224 +502,2828 @@ static psa_status_t psa_its_read_file( psa_storage_uid_t uid,
+    psa_assert( ret == 0 );
+    psa_assert( uid_data_length == psa_cs_testdata_vec3_len );
+
+    get_filename_flags = PSA_CS_GET_FILENAME_F_BA
```
Could it be done in a loop with 2 different UID?

jainvikas8 commented on lib/common_storage.c:
```diff
@@ -146,224 +502,2828 @@ static psa_status_t psa_its_read_file( psa_storage_uid_t uid,
+    psa_assert( ret == 0 );
+    psa_assert( uid_data_length == psa_cs_testdata_vec3_len );
+
+    get_filename_flags = PSA_CS_GET_FILENAME_F_BA
```
Could it be done in a loop with 2 different UID?

simonqhughes replied:
Fixed.

jainvikas8 commented on lib/common_storage.c:
```diff
@@ -146,224 +502,2828 @@ static psa_status_t psa_its_read_file( psa_storage_uid_t uid,
+    psa_assert( ret == 0 );
+    psa_assert( uid_data_length == psa_cs_testdata_vec3_len );
+
+    get_filename_flags = PSA_CS_GET_FILENAME_F_BA
```
Could it be done in a loop with 2 different UID?

Why don't we call a function to perform these steps which are generic for a UID and use it?

simonqhughes replied:
Fixed.

jainvikas8 commented on lib/common_storage.c:
```diff
@@ -146,224 +502,2828 @@ static psa_status_t psa_its_read_file( psa_storage_uid_t uid,
+    psa_assert( ret == 0 );
+    psa_assert( uid_data_length == psa_cs_testdata_vec3_len );
+
+    get_filename_flags = PSA_CS_GET_FILENAME_F_BA
```
Why don't we call a function to perform these steps which are generic for a UID and use it?

simonqhughes replied:
Fixed.

jainvikas8 commented on lib/common_storage.c:
```diff
@@ -146,224 +502,2828 @@ static psa_status_t psa_its_read_file( psa_storage_uid_t uid,
+    psa_assert( ret == 0 );
+    psa_assert( uid_data_length == psa_cs_testdata_vec3_len );
+
+    get_filename_flags = PSA_CS_GET_FILENAME_F_BA
```
Why don't we call a function to perform these steps which are generic for a UID and use it?

simonqhughes replied:
Fixed.

jainvikas8 commented on app/main.c:
```diff
@@ -1636,6 +1653,91 @@ int execute_tests( int argc , const char ** argv )
+            "TC003: Report error for 0 .dat, 0 .bak, 1 .tmp, F_WRITE_ONCE unset)    %s\n",
+            "TC051: Rcvr .dat (1 .dat (new), 2 .bak, F_WRITE_ONCE unset)            %s\n",
+            "TC052: Rcvr .dat (1 .dat (old), 2 .bak, F_WRITE_ONCE unset)            %s\n",
+            "TC053: Rcvr .dat (1 .dat(new, mismatched), 1 .bak, F_WRITE_ONCE unset) %s\n",
```
"TC053: Rcvr .dat (1 .dat (new, mismatched), 1 .bak, F_WRITE_ONCE unset)

simonqhughes replied:
Please can you expand your statement here. Your comment doesn't ask for a specific change.

simonqhughes committed:
SQUASH: Refactor common test code into helper functions.

simonqhughes commented on lib/common_storage.c:
```diff
@@ -76,67 +143,356 @@
+ *  - If the copy is interrupted by a power failure, the incomplete temporary file
+ *    can be detected and deleted upon startup.
+ * ARGUMENTS:
+ *   src_filename       name of file to copy
```
Fixed.

simonqhughes commented on lib/common_storage.c:
```diff
@@ -76,67 +143,356 @@
-                                       psa_cs_api_t api)
+
+/* FUNCTION: psa_cs_read_file_core()
+ *  Open and read the file object header given a filename.
```
Fixed.

simonqhughes commented on lib/common_storage.c:
```diff
@@ -146,224 +502,2828 @@ static psa_status_t psa_its_read_file( psa_storage_uid_t uid,
+ *  Open the file and read the meta-data.
+ * ARGUMENTS:
+ *   uid         IN, unique file object ID.
+ *   p_info      IN, pointer to information structure to take copy of in psa_storage_info_t stored in file.
```
Fixed.

simonqhughes commented on lib/common_storage.c:
```diff
@@ -146,224 +502,2828 @@ static psa_status_t psa_its_read_file( psa_storage_uid_t uid,
-            ( psa_cs_total_size + data_length > PSA_STORAGE_MAX_SIZE ) )
+    psa_debug( " %s\n", "Entry" );
+    /* find xxxx.bakx+1
+     * If only bka exists then state->bkb_seqnum = 0 and then this doesnt work for
```
Fixed.

simonqhughes commented on lib/common_storage.c:
```diff
@@ -146,224 +502,2828 @@ static psa_status_t psa_its_read_file( psa_storage_uid_t uid,
-    psa_cs_num_file_objects++;
-    psa_cs_total_size += data_length;
+    /* find xxxx.bakx+1 (the latest xxxx.bk file.
+     * If only bka exists then state->bkb_seqnum = 0 and then this doesnt work */
```
Fixed.

simonqhughes commented:
@jainvikas8  I've acted upon the outstanding feedback. Please review again. Thanks.

simonqhughes commented:
@jainvikas8 Many thanks for doing this review. Perhaps I can do the same for you sometime.
